docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,702 @@
+# Proposal: Optional Output Syntax
+
+Spawn-on-demand solves many problems but in its pure form it can't tell if the
+workflow has followed the "intended path" - to completion of the whole graph
+(say) vs going down a dead-end side branch. This is because all task outputs 
+are effectively *optional*, and the scheduler just follows the flow wherever it
+leads at runtime.
+
+Cylc 7 could tell (albeit imperfectly!) if the intended path was followed 
+because all task outputs were effectively *required*. All possible paths are
+pre-spawned (to the next cycle point instance of each active task, roughly
+speaking). Users had to explicitly remove unused alternate paths with suicide
+triggers, and any remaining unsatisfied waiting tasks would indicate a problem
+by stalling the workflow.
+
+This document proposes a way to convey the intended path of execution to the
+Cylc 8 spawn-on-demand scheduler by making all task outputs **expected** by
+default and giving a new syntax to explicitly mark **optional** outputs. If a
+task finishes without completing all of its expected outputs it will be
+retained in the n=0 task pool and flagged as an **incomplete task** (if the
+task did not fail unexpectedly then this represents an error against the
+workflow configuration that was not recognized by the task job itself).
+
+*Note: optional triggers were considered and rejected as another way of
+potentially handling this problem - see cylc-admin#128*
+
+## The New Output Syntax
+
+`foo:x` means `x` is an *expected output* of task `foo`. If a task finishes
+without completing all of its expected outputs, we will retain it in the task
+pool as an *incomplete task*.
+
+`foo?y` means `y` is an *optional output* of task `foo`. It is OK for a task to
+finish without completing its optional outputs.
+
+Abbreviated syntax for the success case:
+- `foo?` means `foo?succeed` (success is optional)
+- `foo` and `foo:` mean `foo:succeed` (success is expected)
+  - this supports Cylc 7 graphs as well as being a sensible default
+
+## Syntax Errors
+
+### succeed and fail
+
+The `succeed` and `fail` outputs are mutually exclusive opposites, therefore:
+
+- `succeed` and `fail` can't both be expected (or the task would be flagged as
+  incomplete no matter what)
+```
+# ERROR
+a
+a:fail
+```
+
+- `succeed` and `fail` can both be optional though (at a success/fail alternate
```

Comment:
If a task has only optional outputs, what is considered completion? Any of the options being triggered?

Reply:
If all of a task's outputs are optional (including succeed and fail) then however it *finishes* (succeed or fail) it is complete. It might be easier to think of it as being incomplete only if some expected outputs were not completed.